File: built-in plugins dsp/mb-eq-vintage.cpp
/**
 * MB Vintage EQ
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Pultec-style vintage EQ
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_EQ_VINTAGE_H
#define MB_EQ_VINTAGE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbEqVintage : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-eq-vintage";
    static constexpr const char* PLUGIN_NAME    = "MB Vintage EQ";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float lowBoost = 0f;  // range [0, 10]
    float lowCut = 0f;  // range [0, 10]
    float highBoost = 0f;  // range [0, 10]
    float highAtten = 0f;  // range [0, 10]
    };

    MbEqVintage() = default;
    ~MbEqVintage() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.lowBoost = std::clamp(params.lowBoost, 0f, 10f);
        params.lowCut = std::clamp(params.lowCut, 0f, 10f);
        params.highBoost = std::clamp(params.highBoost, 0f, 10f);
        params.highAtten = std::clamp(params.highAtten, 0f, 10f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Vintage EQ
        return input;
    }
};

#endif // MB_EQ_VINTAGE_H

File: built-in plugins dsp/mb-eq-vintage.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbEqVintagePlugin: PluginDefinition = { id: 'mb-eq-vintage', slug: 'mb-eq-vintage', name: 'MB Vintage EQ', category: 'effect', type: 'eq', version: '1.0.0', description: 'Pultec-style vintage EQ', author: 'Max Booster', parameters: [{ id: 'lowBoost', name: 'Low Boost', type: 'float', defaultValue: 0, minValue: 0, maxValue: 10, automatable: true }, { id: 'lowCut', name: 'Low Cut', type: 'float', defaultValue: 0, minValue: 0, maxValue: 10, automatable: true }, { id: 'highBoost', name: 'High Boost', type: 'float', defaultValue: 0, minValue: 0, maxValue: 10, automatable: true }, { id: 'highAtten', name: 'High Atten', type: 'float', defaultValue: 0, minValue: 0, maxValue: 10, automatable: true }], defaultPreset: { lowBoost: 0, lowCut: 0, highBoost: 0, highAtten: 0 } };

export default MbEqVintagePlugin;
